core/jvmMain/src/kotlinx/serialization/SerializersJvm.kt

```diff
@@ -17,65 +17,86 @@ import java.lang.reflect.*
 import kotlin.reflect.*
 
 /**
- * Reflectively constructs a serializer for the given reflective Java [type].
- * [serializer] is intended to be used as an interoperability layer for libraries like GSON and Retrofit,
- * that operate with reflective Java [Type] and cannot use [typeOf].
+ * Reflectively retrieves a serializer for the given [type].
  *
- * For application-level serialization, it is recommended to use `serializer<T>()` instead as it is aware of
+ * This overload is intended to be used as an interoperability layer for libraries like Retrofit,
+ * that operate with Java's type tokens and cannot use Kotlin's [KType] or [typeOf].
+ * For application-level serialization, it is recommended to use `serializer<T>()` or `serializer(KType)` instead as it is aware of
  * Kotlin-specific type information, such as nullability, sealed classes and object singletons.
  *
+ * Note that because [Type] does not contain any information about nullability, all created serializers
+ * work only with not-null data.
+ *
+ * Custom [Type] implementations are not supported.
```

Comment:
I suggest something like "[Type] implementations from outside of `java.lang.reflect` package are not supported" to be more strict

Reply:
Actually, it can be used if this custom implementation implements GenericArrayType/ParameterizedType/WildcardType (all of those are interfaces)

Reply:
Good point